File: sdks/canary/typescript/scenarios/s-auth.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// S-AUTH canary — TypeScript SDK

import { LLMSafeSpaces, AuthError } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch } from '../canary.js';

async function run(run: Runner, cfg: Config): Promise<void> {
  const c = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: cfg.apiKey, timeout: 20000, fetch: nodeFetch as any });

  // P1: valid API key
  const [ok, me] = await run.assertNoError(() => c.auth.me(), 'valid-key: auth.me no error');
  if (ok && me) {
    run.assert(typeof me.id === 'string' && me.id.length > 0, 'valid-key: user.id present');
    run.assert(typeof me.email === 'string', 'valid-key: user.email present');
    run.assert(typeof me.role === 'string', 'valid-key: user.role present');
    run.assert(me.active === true, 'valid-key: user.active=true');
  }

  // P2+P3: JWT login
  if (cfg.email && cfg.password) {
    const jwtC = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, credentials: { email: cfg.email, password: cfg.password }, timeout: 20000, fetch: nodeFetch as any });
    const [ok2, me2] = await run.assertNoError(() => jwtC.auth.me(), 'jwt-login: auth.me no error');
    if (ok2 && me2) run.assert(me2.active === true, 'jwt-login: user.active=true');
  }

  // N1–N3: bad credentials
  for (const [name, key] of [
    ['invalid-key', 'lsp_invalid_canary_key_000000000000'],
    ['empty-key', ''],
    ['malformed-key', 'not-an-lsp-key'],
  ]) {
    const bad = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: key as string, timeout: 10000, fetch: nodeFetch as any });
    await run.assertError(() => bad.auth.me(), `${name}: AuthError`);
  }
}

async function main() {
  const r = new Runner('auth');
  const cfg = configFromEnv();
  await run(r, cfg);
  const res = r.print();
  process.exit(r.exitCode());
}

main().catch(e => { console.error(e); process.exit(1); });

File: sdks/canary/python/scenarios/s_auth.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""S-AUTH canary — Python SDK"""

from __future__ import annotations

import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from canary import Runner, Config, config_from_env
from llmsafespaces import LLMSafeSpaces, AuthError


def run(run: Runner, cfg: Config) -> None:
    c = LLMSafeSpaces(cfg.api_url, api_key=cfg.api_key, timeout=20.0)

    # P1: valid API key
    ok, me = run.assert_no_error(lambda: c.auth.me(), "valid-key: auth.me no error")
    if ok:
        run.assert_(me.get("id") is not None, "valid-key: user.id present")
        run.assert_(me.get("email") is not None, "valid-key: user.email present")
        run.assert_(me.get("role") is not None, "valid-key: user.role present")
        run.assert_(me.get("active") is True, "valid-key: user.active=true")

    # P2+P3: JWT login
    if cfg.email and cfg.password:
        jwt_c = LLMSafeSpaces(
            cfg.api_url, email=cfg.email, password=cfg.password, timeout=20.0
        )
        ok2, me2 = run.assert_no_error(
            lambda: jwt_c.auth.me(), "jwt-login: auth.me no error"
        )
        if ok2:
            run.assert_(me2.get("active") is True, "jwt-login: user.active=true")

    # N1: invalid key
    bad = LLMSafeSpaces(
        cfg.api_url, api_key="lsp_invalid_canary_key_000000000000", timeout=10.0
    )
    run.assert_error(lambda: bad.auth.me(), "invalid-key: raises AuthError")

    # N2: empty key
    noauth = LLMSafeSpaces(cfg.api_url, api_key="", timeout=10.0)
    run.assert_error(lambda: noauth.auth.me(), "empty-key: raises error")

    # N4+N5: wrong password / nonexistent email
    for name, email, pw in [
        ("wrong-password", cfg.email or "canary@example.com", "definitely-wrong-xyz"),
        ("nonexistent-email", "ghost-99@nonexistent.invalid", "wrongpass123"),
    ]:
        bad_login = LLMSafeSpaces(cfg.api_url, email=email, password=pw, timeout=10.0)
        run.assert_error(lambda c=bad_login: c.auth.me(), f"{name}: raises error")


if __name__ == "__main__":
    r = Runner("auth")
    cfg = config_from_env()
    run(r, cfg)
    res = r.print()
    sys.exit(r.exit_code())
